src/module.ts

```diff
@@ -0,0 +1,193 @@
+import { promises as fsp } from 'fs'
+import * as crypto from 'crypto'
+import { join, resolve } from 'pathe'
+import template from 'lodash.template'
+import { addPlugin, addTemplate, defineNuxtModule, isNuxt2, addComponent } from '@nuxt/kit'
+import { createCommonJS } from 'mlly'
+
+import { name, version } from '../package.json'
+
+const DEFAULTS: ColorModeOptions = {
+  preference: 'system',
+  fallback: 'light',
+  hid: 'nuxt-color-mode-script',
+  globalName: '__NUXT_COLOR_MODE__',
+  componentName: 'ColorScheme',
+  classPrefix: '',
+  classSuffix: '-mode',
+  storageKey: 'nuxt-color-mode'
+}
+
+export default defineNuxtModule({
+  meta: {
+    name,
+    version,
+    configKey: 'colorMode',
+  },
+  defaults: DEFAULTS,
+  async setup (options, nuxt) {
+    const { __dirname } = createCommonJS(import.meta.url)
+
+    // Read script from disk and add to options
+    const scriptPath = resolve(__dirname, 'script.min.js')
+    const scriptT = await fsp.readFile(scriptPath, 'utf-8')
+    options.script = template(scriptT)({ options })
+
+    // Inject options via virtual template
+    nuxt.options.alias['#color-mode-options'] = addTemplate({
+      src: '',
+      filename: 'color-mode-options.mjs',
+      getContents: () => Object.entries(options).map(([key, value]) =>
+        `export const ${key} = ${JSON.stringify(value, null, 2)}
+      `).join('\n')
+    }).dst
+
+    const runtimeDir = isNuxt2() ? resolve(__dirname, 'runtime/vue2') : resolve(__dirname, 'runtime/vue3')
+    nuxt.options.build.transpile.push(runtimeDir)
+
+    // Add plugins
+    for (const template of ['plugin.client', 'plugin.server']) {
+      addPlugin(resolve(runtimeDir, template))
+    }
+
+    addComponent({ name: options.componentName, filePath: resolve(runtimeDir, 'component.vue') })
+
+    // Nuxt 3 - SSR false
+    // TODO: use nitro hooks
+    if (!nuxt.options.ssr) {
+      nuxt.hook('nitro:document', (template) => {
+        template.contents = template.contents.replace('</body>', `</body><script>${options.script}</script>`)
+      })
+    }
+
+    // Nuxt 2 - SSR false
+    nuxt.hook('vue-renderer:spa:prepareContext', ({ head }) => {
+      const script = {
+        hid: options.hid,
+        innerHTML: options.script,
+        pbody: true
+      }
+
+      head.script.push(script)
+
+      const serializeProp = '__dangerouslyDisableSanitizersByTagID'
+      head[serializeProp] = head[serializeProp] || {}
+      head[serializeProp][options.hid] = ['innerHTML']
+    })
+
+    // Nuxt 2 - SSR true
+    nuxt.hook('vue-renderer:ssr:csp', (cspScriptSrcHashes) => {
```

Comment:
Would be a nice idea if we guard these hooks with `isNuxt2()` at least and lazy import dependencies or move to an extra file. Nuxt 2 without Nitro support like this would be lots of trouble in the future.